src/lib/cache.js

```diff
@@ -128,11 +136,14 @@ export default {
     } else {
       data.cached = timestamp
     }
+    const payload = zlib.deflateSync(
+                      JSON.stringify(data)
+                    ).toString('base64')
```

Comment:
`set` is not in a Promise here, so keep sync or wrap it in a Promise?

Reply:
Ah I see. Ok then no need to return a Promise, but we definitely don’t want it to be synchronous, so just change it to use `zlib.deflate(…)` instead without the callback argument.

Reply:
Sorry now I am confused.  `zlib.deflate(…)` requires a callback arg - if one is undefined I get `TypeError: callback is not a function`

Reply:
You might need to wrap it in a promise at the top of the gile to get that behavior, that would look something like:

```js
const deflateP = (data: string) => new Promise((done) => zlib.default(data, done) )
```

Reply:
Oh I’d have expected the callback to be optional. I guess it makes sense, just have the callback log an error if one occurs.

Reply:
The data is compressed (deflated) data is passed to the callback from  `zlib.deflate` - I really am not following what your example does @orta - or why a callback would be optional for the async version of a function @alloy - how then to do anything with the return data?  In any case this is as far as I am but I get race conditions running tests, as the `set` method is no longer synchronous.

```
const deflateP = (dataz) => {
  return new Promise((resolve, reject) =>
    zlib.deflate(JSON.stringify(dataz), (er, deflatedData) => {
      if (er) {
        error(er)
      } else {
        resolve(deflatedData)
      }
    })
  )
}
```

Then in the `set` method:

```
deflateP(data).then(deflatedData => {
      client.set(
        key,
        deflatedData.toString('base64'),
        "EX",
        CACHE_LIFETIME_IN_SECONDS,
        err => {
          if (err) error(err)
        }
      )
    })
```

Reply:
Ah you know what, I wasn’t really thinking about implementation, just the `set` interface that you said returns no promise at the moment. Of course you’re right that you need the callback in your implementation to actually store the deflated data.

As for the test, it looks like you’ll have to return a promise after all from `set`, which you should then return from your test and signal to the test framework that the test is still in progress.

```ts
it('deflates', async () => {
  await cache.set(…)
  expect(…)
})
```

The async/await keywords are sugar syntax around promise handling, without it it would be:

```ts
it('deflates', () => {
  return cache.set(…).then(() => {
    expect(…)
  })
})
```

Reply:
Ahhhhh thanks!  I was attempting to both use a promise and `await` it in the `set` function but jaaa makes total sense that the caller awaits...